### PR TITLE
feat: Support transaction context in trace_transaction

### LINKED ISSUE #6088: Integrate TransactionContext in Js tracer
### Describe the feature

the JsTracer now supports additional tx related fields: https://github.com/paradigmxyz/evm-inspectors/blob/main/src/tracing/js/mod.rs#L547

which we need to provide here:

https://github.com/paradigmxyz/reth/blob/630fd45da6482a150ceedf612293c3cd6773a1dc/crates/rpc/rpc/src/debug.rs#L561-L561 

for this the https://github.com/paradigmxyz/reth/blob/630fd45da6482a150ceedf612293c3cd6773a1dc/crates/rpc/rpc/src/debug.rs#L491-L491

function likely needs to be changed to accept `Option<TransactionContext>`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by supernovahs**
closes #6088

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this has a few conflicts now, sorry, about this -.-

after rebase we need a few changes: populate the transaction context, for example we also need the index of the tx, which we get by modifying `replay_transactions_until` to return the index of the target tx

File: crates/rpc/rpc/src/debug.rs [hunk 1]
```diff
@@ -428,19 +462,58 @@ where
...
+                                    Some(tx_context.with_tx_index(index)),
+                                )?;
+
+                                (trace, state)
+                            } else {
+                                let (trace, state) = this.trace_transaction(
+                                    tracing_options.clone(),
+                                    env,
+                                    &mut db,
+                                    None,
```
Comment (supernovahs) [inline crates/rpc/rpc/src/debug.rs:502]: Is there any way to get tx hash from CallRequest?

**Review by mattsse — CHANGES_REQUESTED** [review]
this can be simplified drastically

File: crates/rpc/rpc/src/debug.rs (hunk 1 above)
Comment (mattsse) [reply crates/rpc/rpc/src/debug.rs:500]: should be None for calls

File: crates/rpc/rpc/src/debug.rs [hunk 2]
```diff
@@ -515,13 +587,22 @@ where
...
+                        let state = res.state.clone();
+                        let result = inspector.json_result(res, &env, db)?;
+                        Ok((GethTrace::JS(result), state))
+                    } else {
+                        let mut inspector = JsInspector::new(code, config)?;
+                        let (res, env, db) = inspect_and_return_db(db, env, &mut inspector)?;
+
+                        let state = res.state.clone();
+                        let result = inspector.json_result(res, &env, db)?;
+                        Ok((GethTrace::JS(result), state))
```
[the remark this answers is not in the document]
Comment (supernovahs) [reply crates/rpc/rpc/src/debug.rs:604]: oops, sorry

[the remark this answers is not in the document]
Comment (mattsse) [reply crates/rpc/rpc/src/debug.rs:604]: all good champ! ;)

File: crates/rpc/rpc/src/debug.rs [hunk 3]
```diff
@@ -82,14 +85,20 @@ where
...
                 let mut db = CacheDB::new(StateProviderDatabase::new(state));
-
                 let mut transactions = transactions.into_iter().peekable();
                 while let Some(tx) = transactions.next() {
                     let tx_hash = tx.hash;
                     let tx = tx_env_with_recovered(&tx);
                     let env = Env { cfg: cfg.clone(), block: block_env.clone(), tx };
-                    let (result, state_changes) =
-                        this.trace_transaction(opts.clone(), env, &mut db).map_err(|err| {
+                    let tx_context = TransactionContext::default();
```
Comment (mattsse) [inline crates/rpc/rpc/src/debug.rs:93]: this can initialize like below with at.as_block_hash

File: crates/rpc/rpc/src/debug.rs [hunk 4]
```diff
@@ -82,14 +85,20 @@ where
             .spawn_with_state_at_block(at, move |state| {
                 let mut results = Vec::with_capacity(transactions.len());
                 let mut db = CacheDB::new(StateProviderDatabase::new(state));
-
                 let mut transactions = transactions.into_iter().peekable();
                 while let Some(tx) = transactions.next() {
```
Comment (mattsse) [inline crates/rpc/rpc/src/debug.rs:89]: this can use `.enumerate` so we get the tx index

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

this is a lot cleaner!

last nits

File: crates/rpc/rpc/src/debug.rs [hunk 5]
```diff
@@ -82,14 +85,22 @@ where
             .spawn_with_state_at_block(at, move |state| {
                 let mut results = Vec::with_capacity(transactions.len());
                 let mut db = CacheDB::new(StateProviderDatabase::new(state));
-
-                let mut transactions = transactions.into_iter().peekable();
-                while let Some(tx) = transactions.next() {
+                for (index, tx) in transactions.iter().enumerate() {
```
Comment (mattsse) [inline crates/rpc/rpc/src/debug.rs:88]: this should remain peekable.

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**